simulation/simulation.py

```diff
@@ -0,0 +1,400 @@
+           birth_date.astype('datetime64[M]').astype(int) % 12 + 1, \
+           birth_date.astype('datetime64[D]').astype(int) % 31 + 1
+
+
+class Simulation:
+    # Note: Tried using scipy.sparse.csr_matrix, but it was 100 times slower than using a dense matrix.
+
+    def __init__(self,
+                 root_folder: str,
+                 settings: SimulationSettings = None,
+                 json_file_name: str = None,
+                 log_verbosity: int = QUIET,
+                 max_cores: int = 1):
+        self._root_folder = root_folder
+        if not os.path.exists(root_folder):
+            os.makedirs(root_folder)
+        self._profile = False
+        self._log_verbosity = log_verbosity
+        self._max_cores = max_cores
+        self._task = SimulationTask()
+        self._configure_logger()
+        if settings is None:
+            self._init_from_json(json_file_name)
+        else:
+            self._init_from_settings(settings)
+            self.save_to_json(os.path.join(self._root_folder, JSON_FILE_NAME))
+
+    def _init_from_json(self, json_file_name: str):
+        logging.info("Loading simulation configuration from %s", json_file_name)
+        with open(json_file_name, "r") as f:
+            loaded = json.load(f)
+        self._settings = SimulationSettings(**loaded["simulation_settings"])
+        self._state_count = loaded["state_count"]
+        self._concept_ids = np.array(loaded["concept_ids"])
+        self._serious_concepts = np.array(loaded["serious_concepts"])
+        self._initial_state_probabilities = np.array(loaded["initial_state_probabilities"])
+        self._dynamic_state_entry_coefs = [np.array(matrix) for matrix in loaded["dynamic_state_entry_coefs"]]
+        self._dynamic_state_exit_probabilities = np.array(loaded["dynamic_state_exit_probabilities"])
+        self._concept_emmision_coefs = [np.array(matrix) for matrix in loaded["concept_emmision_coefs"]]
+
+    def _init_from_settings(self, settings: SimulationSettings):
+        self._settings = settings
+        self._state_count = settings.dynamic_state_count + settings.fixed_state_count + 2  # + 2 for age and sex
+        self._concept_ids = np.arange(settings.concept_count) + 1000000
+
+        # Initialize probabilities and coefficients:
+        self._initial_state_probabilities = np.concatenate((
+            np.random.beta(size=settings.dynamic_state_count, a=1, b=4) / settings.dynamic_state_count,
+            np.random.uniform(size=settings.fixed_state_count + 2)
+        ))
+        # Note: really only need half of these matrices to be filled, because interaction matrix will be symmetrical
+        # over diagonal, but it's easier to just fill them completely:
+        self._dynamic_state_entry_coefs = []
+        for i in range(settings.dynamic_state_count):
+            matrix = np.zeros(self._state_count * self._state_count)
+            non_zero_count = round(len(matrix) / self._state_count)
```

Comment:
Nitpick: Isn't `non_zero_count` equal to `self._state_count`?
```suggestion
            non_zero_count = self._state_count
```

Reply:
Yes, I see that is true now. But I was trying to figure out what proportion of the matrix should be empty. At first I took a fixed proportion (e.g. size / 100), now I tried to make it more flexible by dividing by the state count. May still choose to change that, so I'll keep it as it is.